sdk/resourcemanager/Azure.ResourceManager.Core/src/ApiVersions.cs

```diff
@@ -19,30 +19,30 @@ public class ApiVersions
         /// <summary>
         /// Initializes a new instance of the <see cref="ApiVersions"/> class.
         /// </summary>
-        public ApiVersions()
+        internal ApiVersions(AzureResourceManagerClientOptions clientOptions)
         {
-            BuildApiTable();
+            BuildApiTable(clientOptions);
         }
 
-        private Dictionary<string, string> _resourceToApiVersions = new Dictionary<string, string>();
+        private Dictionary<string, (PropertyInfo, object)> _loadedResourceToApiVersions = new Dictionary<string, (PropertyInfo, object)>();
+        private Dictionary<string, string> _nonLoadedResourceToApiVersion = new Dictionary<string, string>();
 
-        private void BuildApiTable()
+        private void BuildApiTable(AzureResourceManagerClientOptions clientOptions)
         {
             var methods = GetExtensionMethods();
             foreach (var method in methods)
             {
                 if (method.Name.EndsWith("RestApiVersions"))
                 {
-                    var apiObject = method.Invoke(null, new object[] { this });
+                    var apiObject = method.Invoke(null, new object[] { clientOptions });
                     var properties = apiObject.GetType().GetProperties();
                     foreach (var prop in properties)
                     {
                         if (typeof(ApiVersionsBase).IsAssignableFrom(prop.PropertyType))
                         {
                             var propVal = (ApiVersionsBase)prop.GetValue(apiObject);
-                            Console.WriteLine(prop.GetType());
                             var key = propVal.ResourceType;
-                            _resourceToApiVersions.Add(key.ToString(), propVal);
+                            _loadedResourceToApiVersions.Add(key.ToString(), (prop, apiObject));
```

Comment:
I think we should abstract the usage to a PropertyWrapper class which takes in these two params and has a Set and Get method that hid the reflection

Reply:
Since this is internal and never exposed outside of this class, what advantage does this have?

It would just be an object with two properties, a PropertyInfo and then a Object associated with that property?

Reply:
The advantage is our code read much simpler and you abstract away the usage of setting and getting